### PR TITLE
[stdlib] prevent preventable buffer overflows when constructing Strings from C strings.

### PR DESCRIPTION
**Pull request by glessard**
The family of `String` initializers that convert from C strings (null-terminated byte buffers) could until now be called with Swift arrays. When the array passed in to them violates the C string precondition of containing a zero byte, this would result in a buffer overflow. This is unexpected when the entirety of the code is pure Swift. This patch adds overloads to capture arguments that would become array-to-pointer conversions and inout-to-pointer conversions. The new overloads prevent reading past the end of these allocations. Note that the `String` conversion is not susceptible to this overflow bug, since it always appends a trailing null.

The `String` construction methods affected are:
- `String.init(cString: UnsafePointer<UInt>)` and `String.init(cString: UnsafePointer<CChar>)`
- `String.init?(validatingCString: UnsafePointer<CChar>)`
- `String.decodeCString<Encoding>(_ cString: UnsafePointer<Encoding.CodeUnit>?, as: Encoding.Type, repairingInvalidCodeUnits: Bool = true)`
- `String.init<Encoding>(decodingCString: UnsafePointer<Encoding.CodeUnit>, as: Encoding.Type)`

All of these have been overloaded to capture misuses of array conversion and inout conversion. In both cases, if there is no zero-valued code unit within the allocation, a precondition is considered violated and the function traps. In the case of the inout conversion, the only valid code unit value is zero, so we also added a deprecation warning.

The construction methods above were also overloaded to capture `String` conversions. The `String` conversions are safe but wasteful, so we added a deprecation warning to those as well, pointing out that using the `String` argument directly is more straightforward.

Note that we considered adding the `@_nonEphemeral` attribute to the affected `UnsafePointer` parameters, and concluded that it is not an appropriate solution, because it only adds a warning without fixing the buffer overflow issue.

Resolves rdar://90336023.

### CODE REVIEW

**Comment by glessard** [comment]
@swift-ci Please Test Source Compatibility

**Comment by glessard** [comment]
@swift-ci please test

**Comment by glessard** [comment]
@swift-ci please test

**Comment by glessard** [comment]
@swift-ci please test

**Comment by glessard** [comment]
@swift-ci please test

**Comment by glessard** [comment]
I'm tempted to mark these new overloads as deprecated, but that probably requires evolution approval.

**Comment by glessard** [comment]
@swift-ci please test

**Comment by glessard** [comment]
@swift-ci please smoke test

**Comment by glessard** [comment]
@swift-ci please test

**Comment by glessard** [comment]
@swift-ci please smoke test linux platform

File: stdlib/public/core/CString.swift [hunk 1]
```diff
@@ -42,22 +42,53 @@ extension String {
...
     self = String._fromUTF8Repairing(
-      UnsafeBufferPointer(start: cString, count: len)).0
+      UnsafeBufferPointer(start: nullTerminatedUTF8, count: len)).0
+  }
+
+  @inlinable
+  @_alwaysEmitIntoClient
+  public init(cString nullTerminatedUTF8: [UInt8]) {
+    self = nullTerminatedUTF8.withUnsafeBufferPointer(
+      String.init(_checkingCString:)
```
Comment (Azoy) [inline stdlib/public/core/CString.swift:90]: this is a little weird, I think the regular closure here would be easier to parse.

**Comment by Azoy** [comment]
Changes look good to me. I think some of the API here requires evolution however.

File: stdlib/public/core/CString.swift (hunk 1 above)
Comment (glessard) [reply stdlib/public/core/CString.swift:90]: Agreed. It was on one line before I added the longer parameter label from the protocol.

**Review by Azoy — APPROVED** [review]

**Comment by glessard** [comment]
@swift-ci please smoke test

**Comment by glessard** [comment]
Note: the added overloads are not new API, in the sense that their type parameters are refinements of those of the existing calls.

**Comment by glessard** [comment]
I'd like feedback on my deprecation messages, including the empty ones.

**Comment by glessard** [comment]
@ swift-ci please smoke test

**Comment by glessard** [comment]
@swift-ci please smoke test

**Review by Azoy — APPROVED** [review]
Looks good, thank you!

**Comment by glessard** [comment]
@swift-ci please smoke test

**Comment by glessard** [comment]
@swift-ci please test Windows platform

**Comment by glessard** [comment]
The linux failure seems like a problem with stdlibunittest; it seems to not be doing the right thing with `expectCrashLater(withMessage:)`.

File: stdlib/public/core/CString.swift [hunk 2]
```diff
@@ -42,22 +42,84 @@ extension String {
...
     self = String._fromUTF8Repairing(
-      UnsafeBufferPointer(start: cString._asUInt8, count: len)).0
+      UnsafeBufferPointer(start: nullTerminatedUTF8._asUInt8, count: len)).0
+  }
+
+  @inlinable
+  @_alwaysEmitIntoClient
+  public init(cString nullTerminatedUTF8: [CChar]) {
+    self = nullTerminatedUTF8.withUnsafeBytes {
+      String(_checkingCString: $0.assumingMemoryBound(to: UInt8.self))
```
Comment (atrick) [inline stdlib/public/core/CString.swift:56]: > String(_checkingCString: $0.assumingMemoryBound(to: UInt8.self))

This looks like undefined behavior. The optimizer is still super conservative with pointer types, so it's not causing miscompiles yet.

It's sad that _fromUTF8Repairing doesn't just take a raw pointer. But if you must view `CChar` memory as `UInt8`, then `withMemoryRebound` is perfect. 

Maybe as a separate PR, I implore you to remove `_asUInt8`, which never works.

Comment (glessard) [reply stdlib/public/core/CString.swift:56]: Agreed. Removing `_asUInt8` and `_asCChar` would be a good thing, but there are enough uses scattered throughout that it should be its own PR. (rdar://91343475)

**Comment by glessard** [comment]
@swift-ci please smoke test

**Comment by sergiocampama** [comment]
@glessard I believe this PR introduced an issue where before 5.7 this would work:

```
var path: (CChar, CChar, CChar, CChar, CChar) = [47, 118, 97, 114, 0]
print(String(cString: &path.0))
```

but with this change in 5.7 it does not.

I believe it is because there is a new `String(cString nullTerminatedUTF8: inout CChar)` which incorrectly checks for the first element is zero (`nullTerminatedUTF8 == 0`), when in reality it should find the last element in memory which is zero.

The `inout` API takes precedence over the `UnsafePointer<CChar>` API that existed before (which continues to work), which makes all usage of this API fail with the preconditionFailure, and returns empty strings. I tested precedence  with this example:

```
func something(_ c: UnsafePointer<UInt8>) {
  print("from unsafe pointer", c)
}

//func something(_ c: inout UInt8) {
//  print("from inout", c)
//}

var pointer: (UInt8, UInt8, UInt8) = (1, 2, 0)
something(&pointer.0)
```

Is it expected that this API would change in this way? Are previous usages of this API wrong? 

It feels off that the deprecation warning states to change the API, and that the resulting behavior is at the same time changed. At least, it should be a compilation error to prevent people who don't usually fix deprecation warnings as quickly from building code that will start to fail under 5.7 (meaning probably macOS Ventura and 5.7 linux images)

Found this while running the tests of `git@github.com:swhitty/FlyingFox.git` on macOS ventura (22A5295i) and Xcode 14 beta 4 (14A5284g)

**Comment by stephentyrone** [comment]
Hi @sergiocampama!

Unfortunately, this:
```
var path: (CChar, CChar, CChar, CChar, CChar) = [47, 118, 97, 114, 0]
print(String(cString: &path.0))
```
never actually worked before. (It will have happened to "work" and produce the desired result on accident occasionally, but it steps outside of the language model, and would have suddenly stopped working with a new language version or compiler optimization). In particular, the pointer created by `&path.0` only points to a single `CChar` with value `47`; any bytes that follow that are undefined. This code was never correct, and that is now being diagnosed.

@glessard can help you with an appropriate workaround.

**Comment by glessard** [comment]
@sergiocampama Your initial example exploited an explicit out-of-bounds read: your are passing in an `inout CChar`, that type has a memory size of 1. Yet you were reading 4 bytes. That's a bug! Furthermore, in the example as written, the pointer you get from this conversion points to the value in `path.1`, but advancing that by 1 is not guaranteed to point to `path.1`. It just so happens that it generally does, and will if that tuple comes from an imported C api.

The workaround you identified is appropriate. Note that the fact that there is no proper bounds checking is made explicit by the fact that you are using `withUnsafePointer(to:)`. That is intentional.

**Comment by sergiocampama** [comment]
Thanks @glessard and @stephentyrone for the additional context. I do understand that this is unsafe and a bug and should be updated. Just merely pointing out that in this new version it is not being diagnosed, it does changes behavior for current bad users of the `cString` API, where before it used to erroneously work, and now it crashes. 

IMO it should be marked as unavailable and not a deprecation, since there could be code out there that (erroneously) compiled and worked and now will crash, especially for a minor release (changing from 5.6 to 5.7, unclear if Swift itself follows semver). (Also, the initializer as is does nothing of value, so it's unclear why it would even stay.)

This crash will appear either when running previously compiled apps which are now running on Swift 5.7 (such as Ventura or newer linux images, not sure how to mitigate this case) or when compiling old code and disregarding deprecation notices (which a non trivial amount of people do).

Finally, just pointing this out as mismatched expectations on changing behavior, if you say this is expected then that's that.

ps: Thanks for your work in making Swift safer :)

**Comment by stephentyrone** [comment]
The behavior of previously compiled apps is unchanged here; the pre-existing symbol has the same behavior as before (please confirm, @glessard). This only changes the behavior for apps that are recompiled, at which point one would hope that the resulting binary is tested.

I _do_ think that we should consider simply making the `inout CChar` version unavailable in a future release, as it can really only be used in error.

**Comment by sergiocampama** [comment]
Oh, right, it's emitted into the client, missed that.

**Comment by stephentyrone** [comment]
Right, but the existence of the new symbol doesn't change which symbol existing code uses. You have to recompile to get the new behavior.

**Pull request merged**